masonry/src/text/selection.rs

```diff
@@ -265,6 +266,111 @@ impl<T: Selectable> TextWithSelection<T> {
         }
         self.layout.draw(scene, point);
     }
+
+    fn access_position_from_offset(&self, offset: usize, affinity: Affinity) -> TextPosition {
+        let text = self.text().as_ref();
+        debug_assert!(offset <= text.len(), "offset out of range");
+
+        for (line_index, line) in self.layout.layout.lines().enumerate() {
+            let range = line.text_range();
+            if !(range.contains(&offset)
+                || (offset == range.end
+                    && (affinity == Affinity::Upstream
+                        || line_index == self.layout.layout.len() - 1)))
+            {
+                continue;
+            }
+
+            for (run_index, run) in line.runs().enumerate() {
+                let range = run.text_range();
+                if !(range.contains(&offset)
+                    || (offset == range.end
+                        && (affinity == Affinity::Upstream
+                            || (line_index == self.layout.layout.len() - 1
+                                && run_index == line.len() - 1))))
+                {
+                    continue;
+                }
+
+                let run_offset = offset - range.start;
+                let run_path = (line_index, run_index);
+                let id = *self.layout.access_ids_by_run_path.get(&run_path).unwrap();
+                let character_lengths =
+                    self.layout.character_lengths_by_access_id.get(&id).unwrap();
+                let mut length_sum = 0_usize;
+                for (character_index, length) in character_lengths.iter().copied().enumerate() {
+                    if run_offset == length_sum {
+                        return TextPosition {
+                            node: id,
+                            character_index,
+                        };
+                    }
+                    length_sum += length as usize;
+                }
+                return TextPosition {
+                    node: id,
+                    character_index: character_lengths.len(),
+                };
+            }
+        }
+
+        panic!("offset not within the range of any run");
```

Comment:
This should probably be a warning or a debug assert? We don't want to panic in release builds because the user found a corner case in the selection code.

Reply:
OK, I had to modify `access_position_from_offset` to return an `Option` so it would have something to return in that case, then modify the caller accordingly. But I think you're right. If I were using an application that had a bug which triggered this code path, I would want it to run with degraded functionality rather than crashing because the developer didn't test accessibility enough (something that's very likely).